source/dlangbot/github_api.d

```diff
@@ -34,9 +37,50 @@ struct Result
 
 Result ghGetRequest(string url)
 {
-    return request(url, (scope req) {
+    import std.digest.sha : sha1Of;
+    import std.digest : toHexString;
+    import std.file : exists, readText;
+
+    auto cacheFileName = "cache/" ~ cast(string)sha1Of(url).toHexString;
+    Result cacheEntry;
+    if (cacheFileName.exists)
+        cacheEntry = cacheFileName.readText.parseJsonString.deserializeJson!Result();
+
+    auto res = request(url, (scope req) {
         req.headers["Authorization"] = githubAuth;
-    }).expectOK.Result;
+        if (auto p = "ETag" in cacheEntry.headers)
+            req.headers["If-None-Match"] = *p;
+        if (auto p = "Last-Modified" in cacheEntry.headers)
+            req.headers["If-Modified-Since"] = *p;
+    });
+    scope(exit) res.dropBody;
+
+    if (res.statusCode == 304 /*Not Modified*/)
+    {
+        logInfo(" > Cache hit!");
+        return cacheEntry;
+    }
+    else
+    if (res.statusCode == 200 /*OK*/)
+    {
+        logInfo(" > Cache miss; ratelimit: %s/%s",
+            res.headers.get("X-Ratelimit-Remaining", "?"),
+            res.headers.get("X-Ratelimit-Limit", "?"),
+        );
+        auto result = res.Result;
+        ensurePathExists(cacheFileName);
+        result.serializeToJsonString.toFile(cacheFileName);
+        return result;
+    }
+    else
+    if (res.statusCode / 100 == 3 && "Location" in res.headers)
+    {
+        auto location = res.headers["Location"];
+        logInfo(" > Redirect: %s", location);
+        return ghGetRequest(applyRelativeURL(url, location));
+    }
```

Comment:
This is not related to caching is it ?

Reply:
Ah, no. I've reused code from dtest.dlang.io. The GitHub API does return redirects in some situations (e.g. https://api.github.com/repos/D-Programming-Language/dmd ) , so I think we do need to handle them in case something gets renamed or moved around in the future.

Reply:
Could you split it in its own commit then ?

Reply:
Sure, done.